{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Simulations for evaluating archaic ancestry inference"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set paths and import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run in archanc root directory\n",
    "import os\n",
    "proj_dir = os.getcwd()\n",
    "msprime_dir = proj_dir +\"/output/msprime/\"\n",
    "archie_src_dir = proj_dir + \"/src/ArchIE/\"\n",
    "archie_out_dir = proj_dir + \"/output/ArchIE/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, proj_dir + '/src/stdpopsim')\n",
    "\n",
    "from stdpopsim import homo_sapiens, models\n",
    "import msprime\n",
    "import itertools\n",
    "import random\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import math"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## define additional demographic models in `stdpopsim` model format\n",
    "\n",
    "To-do:\n",
    "- integrate into stdpopsim github repository"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class TennessenTwoPopOutOfAfrica(models.Model):\n",
    "    \"\"\"\n",
    "    The model is derived from the Tennesen et al.\n",
    "    `analysis <https://doi.org/10.1126/science.1219240>`_  of the jSFS from\n",
    "    European Americans and African Americans.\n",
    "\n",
    "    Model parameters are taken from Fig. S5 in\n",
    "    `Fu et al. (2013) <https://doi.org/10.1038/nature11690>`_.\n",
    "\n",
    "    .. todo:: document this model, including the original publications\n",
    "        and clear information about what the different population indexes\n",
    "        mean.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self):\n",
    "        super().__init__()\n",
    "\n",
    "        generation_time = 25\n",
    "        T_AF = 148e3 / generation_time\n",
    "        T_OOA = 51e3 / generation_time\n",
    "        T_EU0 = 23e3 / generation_time\n",
    "        T_EG = 5115 / generation_time\n",
    "\n",
    "        # Growth rates\n",
    "        r_EU0 = 0.00307\n",
    "        r_EU = 0.0195\n",
    "        r_AF = 0.0166\n",
    "\n",
    "        # population sizes\n",
    "        N_A = 7310\n",
    "        N_AF = 14474\n",
    "        N_B = 1861\n",
    "        N_EU0 = 1032\n",
    "        N_EU1 = N_EU0 / math.exp(-r_EU0 * (T_EU0-T_EG))\n",
    "\n",
    "        # migration rates\n",
    "        m_AF_B = 15e-5\n",
    "        m_AF_EU = 2.5e-5\n",
    "\n",
    "        # present Ne\n",
    "        N_EU = N_EU1 / math.exp(-r_EU * T_EG)\n",
    "        N_AF = N_AF / math.exp(-r_AF * T_EG)\n",
    "\n",
    "        self.population_configurations = [\n",
    "            msprime.PopulationConfiguration(initial_size=N_AF, growth_rate=r_AF),\n",
    "            msprime.PopulationConfiguration(initial_size=N_EU, growth_rate=r_EU)\n",
    "        ]\n",
    "\n",
    "        self.migration_matrix = [\n",
    "            [0, 0],\n",
    "            [0, 0],\n",
    "        ]\n",
    "\n",
    "        self.demographic_events = [\n",
    "            msprime.MigrationRateChange(\n",
    "                time=T_EG, rate=m_AF_EU, matrix_index=(0, 1)),\n",
    "            msprime.MigrationRateChange(\n",
    "                time=T_EG, rate=m_AF_EU, matrix_index=(1, 0)),\n",
    "            msprime.PopulationParametersChange(\n",
    "                time=T_EG, growth_rate=r_EU0, initial_size=N_EU1, population_id=1),\n",
    "            msprime.PopulationParametersChange(\n",
    "                time=T_EG, growth_rate=0, initial_size=N_AF, population_id=0),\n",
    "            msprime.MigrationRateChange(\n",
    "                time=T_EU0, rate=m_AF_B, matrix_index=(0, 1)),\n",
    "            msprime.MigrationRateChange(\n",
    "                time=T_EU0, rate=m_AF_B, matrix_index=(1, 0)),\n",
    "            msprime.PopulationParametersChange(\n",
    "                time=T_EU0, initial_size=N_B, growth_rate=0, population_id=1),\n",
    "            msprime.MassMigration(\n",
    "                time=T_OOA, source=1, destination=0, proportion=1.0),\n",
    "            msprime.PopulationParametersChange(\n",
    "                time=T_AF, initial_size=N_A, population_id=0)\n",
    "        ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function for simulating MNMs on tree"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Randomly select a fraction of variant sites to be MNMs. The second mutation in each MNM is randomly placed from 1-100bp downstream of the first, and the genotypes of the first mutation are duplicated. \n",
    "\n",
    "To-do: separate functions for the MNM simulation from .snp and .geno file generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_mnms(ts, model_label, output_dir=\"./\", rep_label=0, mnm_dist=100, mnm_frac=0.015):\n",
    "    \n",
    "    # .snp file names are underscore-delimited with model, population MNM parameters (dash-delimited with distance & fraction), and replicate label\n",
    "    # .geno file names add the population label\n",
    "#     prefix = model_label+\"_\"+pop_label+\"_mnm\"+str(mnm_dist)+\"-\"+str(mnm_frac)+\"_\"+str(rep_label)\n",
    "    prefix = model_label+\"_mnm\"+str(mnm_dist)+\"-\"+str(mnm_frac)+\"_\"+str(rep_label)\n",
    "    \n",
    "    geno = np.zeros(200, dtype=np.int8)\n",
    "#     eur_geno = np.zeros(100, dtype=np.int8)\n",
    "#     afr_geno = np.zeros(100, dtype=np.int8)\n",
    "\n",
    "    mnm_dict = {}\n",
    "    with open(output_dir + prefix + \".snp\", \"w\") as text_file:\n",
    "        for variant in list(ts.variants()):\n",
    "            print(\"\\t\".join([\"1:\"+str(round(variant.site.position)), \n",
    "                             \"1\", \n",
    "                             str(variant.site.position/10e6), \n",
    "                             str(round(variant.site.position)), \n",
    "                             \"A\", \n",
    "                             \"G\"]), file=text_file)\n",
    "            geno = np.vstack([geno, variant.genotypes])\n",
    "            \n",
    "            random.seed(variant.site.position)\n",
    "            if random.random() < mnm_frac:\n",
    "\n",
    "                dist = random.randint(1,mnm_dist)\n",
    "                mnm_cand = variant.site.position+dist \n",
    "                mnm_cand_r = str(round(mnm_cand))\n",
    "                mnm_dict[mnm_cand_r] = [sum(variant.genotypes[0:100])/100, # allele freq in AFR\n",
    "                                        sum(variant.genotypes[100:200])/100, # allele freq in EUR\n",
    "                                        ts.genotype_matrix().shape[0]] # total number of variants before MNMs\n",
    "                print(\"\\t\".join([\"1:\"+mnm_cand_r, \n",
    "                                 \"1\", \n",
    "                                 str(mnm_cand/10e6), \n",
    "                                 mnm_cand_r, \n",
    "                                 \"A\", \n",
    "                                 \"G\"]), file=text_file)\n",
    "                geno = np.vstack([geno, variant.genotypes])\n",
    "\n",
    "    # output MNM data\n",
    "    mnm_df = pd.DataFrame(mnm_dict).T\n",
    "    mnm_df.index.name = \"MNM_POS\"\n",
    "    mnm_df = mnm_df.rename(columns={0: \"AF_AFR\", \n",
    "                                    1: \"AF_EUR\", \n",
    "                                    2: \"N_TOT\"})\n",
    "    mnm_df.to_csv(output_dir + prefix + \".mnms\", \n",
    "                  index=True,\n",
    "                  sep=\"\\t\")\n",
    "    \n",
    "    # delete empty first row used to initialize genotype matrix            \n",
    "    geno = np.delete(geno, (0), axis=0)\n",
    "    \n",
    "    geno_afr = geno[:,:100]\n",
    "    geno_eur = geno[:,100:200]\n",
    "    \n",
    "    geno_pop = [geno_afr, geno_eur]\n",
    "    \n",
    "    for i, pop in enumerate([\"afr\", \"eur\"]):\n",
    "        np.savetxt(output_dir + prefix + \"_\" + pop + \".geno\", geno_pop[i], delimiter=\"\", fmt='%i')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simulate models\n",
    "\n",
    "Simulate 200 samples (100 each of European and African ancestry) under each of the specified models (with 1000 replicates each)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# coalescent simulation parameters\n",
    "sample_size = 100 # each\n",
    "length = 50000\n",
    "mu = 1.15e-8\n",
    "rr = 1e-8\n",
    "replicates = 1000\n",
    "seed = 30\n",
    "\n",
    "# Gutenkunst 3-population model\n",
    "GutenkunstThreePop_model = homo_sapiens.GutenkunstThreePopOutOfAfrica()\n",
    "GutenkunstThreePop_ts = msprime.simulate(\n",
    "    # first 100 samples from AFR, next 100 from EUR\n",
    "    samples=[msprime.Sample(0, 0)]*sample_size + [msprime.Sample(1, 0)]*sample_size,\n",
    "    length=length, \n",
    "    mutation_rate=mu, \n",
    "    recombination_rate=rr,\n",
    "    random_seed=seed,\n",
    "    num_replicates=replicates,\n",
    "    **GutenkunstThreePop_model.asdict())\n",
    "\n",
    "# Tennessen 2-population model\n",
    "TennessenTwoPop_model = TennessenTwoPopOutOfAfrica()\n",
    "TennessenTwoPop_ts = msprime.simulate(\n",
    "    # first 100 samples from AFR, next 100 from EUR\n",
    "    samples=[msprime.Sample(0, 0)]*sample_size + [msprime.Sample(1, 0)]*sample_size,\n",
    "    length=length, \n",
    "    mutation_rate=mu, \n",
    "    recombination_rate=rr,\n",
    "    random_seed=seed,\n",
    "    num_replicates=replicates,\n",
    "    **TennessenTwoPop_model.asdict())\n",
    "\n",
    "#-------------------------------------------------------\n",
    "# define other models here and add to model_dict below\n",
    "#-------------------------------------------------------\n",
    "\n",
    "# modify demographic parameters to include archaic branches\n",
    "# GutenkunstThreePopArchaic_model = homo_sapiens.GutenkunstThreePopArchaic()\n",
    "\n",
    "# GutenkunstThreePopArchaic_ts = msprime.simulate(\n",
    "#     # first 100 samples from AFR, next 100 from EUR\n",
    "#     samples=[msprime.Sample(0, 0)]*sample_size + [msprime.Sample(1, 0)]*sample_size,\n",
    "#     length=length, \n",
    "#     mutation_rate=mu, \n",
    "#     recombination_rate=rr,\n",
    "#     random_seed=seed,\n",
    "#     num_replicates=replicates,\n",
    "#     **GutenkunstThreePopArchaic_model.asdict())\n",
    "\n",
    "model_dict = {\"GutenkunstThreePop\": GutenkunstThreePop_ts,\n",
    "             \"TennessenTwoPop\": TennessenTwoPop_ts}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add MNMs\n",
    "\n",
    "Loop through models/replicates and randomly generate MNMs.\n",
    "\n",
    "Outputs the .mnm, .snp, and .geno files to `output/msprime/` then passes these as input to the `calc_stats_window_data.py` script from `ArchIE` to calculate the per-sample summary stats that act as features in the `ArchIE` inference method. These output files are then concatenated and passed as the test data to `ArchIE/train/train.R` to predict the fraction of archaic ancestry in each simulated sample.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# MNM simulation parameters\n",
    "mnm_dist = 100\n",
    "mnm_frac = 0.015\n",
    "\n",
    "# change the run_archie flag to True to run ArchIE on the simulated data\n",
    "run_archie = False\n",
    "\n",
    "for model_label, model in model_dict.items():\n",
    "    \n",
    "    print(\"Simulating MNMs on \" + model_label)\n",
    "    \n",
    "    for j, ts in enumerate(model):\n",
    "\n",
    "        prefix = model_label + \"_mnm\" + str(mnm_dist) + \"-\" + str(mnm_frac) + \"_\"\n",
    "\n",
    "        # add MNMs and write out .snp and .geno files\n",
    "        add_mnms(ts, \n",
    "                 output_dir=msprime_dir, \n",
    "                 model_label=model_label, \n",
    "                 mnm_dist=mnm_dist, \n",
    "                 mnm_frac=mnm_frac, \n",
    "                 rep_label=str(j))\n",
    "\n",
    "        for pop in [\"afr\", \"eur\"]:\n",
    "            \n",
    "            # write out separate .ind files per population\n",
    "            with open(msprime_dir + prefix + pop + \"_ADMIXED.ind\", \"w\") as id_file:\n",
    "                for sampleid in range(0,100):\n",
    "                    print(prefix + \"model_\" + str(j) + \"_\" + pop + \"_sample_\" + str(sampleid) + \\\n",
    "                         \"\\tU\\tADMIXED\",  file=id_file)\n",
    "\n",
    "            # run archie on AFR and EUR simulated data\n",
    "            if run_archie:\n",
    "                if pop == \"afr\":\n",
    "                    ref_pop = \"eur\"\n",
    "                else:\n",
    "                    ref_pop = \"afr\"\n",
    "                stats_pop_cmd = \"python \" + archie_src_dir + \"data/calc_stats_window_data.py\" + \\\n",
    "                    \" -s \" + msprime_dir + prefix + str(j) + \".snp\" + \\\n",
    "                    \" -i \" + output_dir + prefix + pop + \"_ADMIXED.ind\" + \\\n",
    "                    \" -a \" + msprime_dir + prefix + str(j) + \"_\" + pop + \".geno\" + \\\n",
    "                    \" -r \" + msprime_dir + prefix + str(j) + \"_\" + ref_pop + \".geno\" + \\\n",
    "                    \" -c 1 -b 0 -e 50000 -w 50000 -z 50000 \" + \\\n",
    "                    \" > \" + archie_out_dir + prefix + str(j) + \"_\" + pop + \".txt\"  \n",
    "        #         print(stats_pop_cmd)\n",
    "                os.system(stats_pop_cmd)\n",
    "                \n",
    "        # combine .ind files into single file, for ANCESTRYMAP -> VCF format conversion\n",
    "        cat_cmd = \"cat \" + msprime_dir + prefix + \"afr_ADMIXED.ind \" + \\\n",
    "            msprime_dir + prefix + \"eur_ADMIXED.ind > \" + msprime_dir + prefix + \"combined.ind\"\n",
    "        os.system(cat_cmd)\n",
    "        \n",
    "        # add code for converting .snp/.geno/.ind data to 1 VCF per replicate\n",
    "        # ...\n",
    "        # ...\n",
    "    \n",
    "    # once analyses of all replicates for a given model are complete, combine and clean up ArchIE output\n",
    "    if run_archie:\n",
    "        for pop in [\"afr\", \"eur\"]:\n",
    "            cat_pop_cmd = \"cat \" + \\\n",
    "                archie_out_dir + \"*\" + pop + \".txt\" + \\\n",
    "                \" > \" + archie_out_dir + \"combined/\" + prefix + pop + \".txt\" \n",
    "            print(cat_pop_cmd)\n",
    "            os.system(cat_pop_cmd)\n",
    "\n",
    "        # remove ArchIE output for individual replicates\n",
    "        clean_cmd = \"rm \" + archie_out_dir + \"*\"\n",
    "        print(clean_cmd)\n",
    "        os.system(clean_cmd)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "anaconda",
   "language": "python",
   "name": "anaconda"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
